### PR TITLE
Add timestampadd/timestampdiff function

### PR DESCRIPTION
**Pull request by HangyuanLiu**
1.Support timestampdiff function
2.Support timestampadd function
3.Add Doris keyword (YRAR,MONTH,WEEK,DAY,HOUR,MINUTE,SECOND)
4.Fix some cup format

#2703

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java [hunk 1]
```diff
@@ -120,65 +120,142 @@ private Type fixType() {
...
+                                + getChild(1).getType() + "' which is incompatible with expected type 'INT'.");
+                    }
+                    castChild(Type.INT, 1);
+                }
+
+                type = dateType;
+                opcode = getOpCode();
+                funcOpName = String.format("%sS_%s", timeUnit,
+                        (op == ArithmeticExpr.Operator.ADD) ? "ADD" : "SUB");
+                // For the month interval, use the invisible special-case implementation.
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java]: remove this comment?

File: fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java [hunk 2]
```diff
@@ -120,65 +120,142 @@ private Type fixType() {
...
-        //     funcOpName += "_INTERVAL";
-        // }
+            type = dateType;
+            opcode = getOpCode();
+            funcOpName = String.format("%sS_%s", timeUnit,
+                    (op == ArithmeticExpr.Operator.ADD) ? "ADD" : "SUB");
+        }
 
         fn = getBuiltinFunction(analyzer, funcOpName.toLowerCase(),
                 collectChildReturnTypes(), Function.CompareMode.IS_NONSTRICT_SUPERTYPE_OF);
```
Comment (kangkaisen) [inline fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java]: `fn is {} name is {}` thie log should be remove or change to debug level.  There will be a lot of this log from doris 0.11

File: be/src/exprs/timestamp_functions.cpp [hunk 3]
```diff
@@ -351,6 +351,92 @@ DateTimeVal TimestampFunctions::timestamp_time_op(
...
+}
+
+BigIntVal TimestampFunctions::seconds_diff(
+        FunctionContext* ctx, const DateTimeVal& ts_val1, const DateTimeVal& ts_val2) {
+    return timestamp_diff<SECOND>(ctx, ts_val1, ts_val2);
+}
+
+template <TimeUnit unit>
+BigIntVal TimestampFunctions::timestamp_diff(FunctionContext* ctx, const DateTimeVal& ts_val2, const DateTimeVal& ts_val1) {
+    std::cout << "null" << std::endl;
```
[the remark this answers is not in the document]
Comment (HangyuanLiu) [reply be/src/exprs/timestamp_functions.cpp]: Sorry, I forget

File: fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java (hunk 1 above)
Comment (HangyuanLiu) [reply fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java]: Done

File: fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java (hunk 2 above)
Comment (HangyuanLiu) [reply fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java]: OK

**Review by imay — COMMENTED** [review]
please add some comment for the new added function

File: docs/documentation/cn/sql-reference/sql-functions/date-time-functions/timestampadd.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# day
```
Comment (imay) [inline docs/documentation/cn/sql-reference/sql-functions/date-time-functions/timestampadd.md]:
```suggestion
# TIMESTAMPADD
```

File: docs/documentation/cn/sql-reference/sql-functions/date-time-functions/timestampdiff.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+# day
```
Comment (imay) [inline docs/documentation/cn/sql-reference/sql-functions/date-time-functions/timestampdiff.md]:
```suggestion
# TIMESTAMPDIFF
```

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java [hunk 6]
```diff
@@ -120,69 +120,141 @@ private Type fixType() {
...
+                }
+                // The first child must return a timestamp or null.
+                if (!getChild(0).getType().isDateType() && !getChild(0).getType().isNull()) {
+                    if (!dateType.isValid()) {
+                        throw new AnalysisException("Operand '" + getChild(0).toSql()
+                                + "' of timestamp arithmetic expression '" + toSql() + "' returns type '"
+                                + getChild(0).getType() + "'. Expected type 'TIMESTAMP/DATE/DATETIME'.");
+                    }
+                    castChild(dateType, 0);
+                }
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/analysis/TimestampArithmeticExpr.java]: seems that these codes are the same with others.
Could you try to reduce these same code?

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**